pretrain/tinyllama.py

```diff
@@ -0,0 +1,340 @@
+"""
+This script is adapted from TinyLlama:
+https://github.com/jzhang38/TinyLlama/blob/main/pretrain/tinyllama.py
+"""
+import math
+import sys
+import time
+from functools import partial
+from pathlib import Path
+from typing import Tuple, Union
+
+import lightning as L
+import torch
+import torch.nn as nn
+from lightning.data import StreamingDataset
+from lightning.data.streaming.item_loader import TokensLoader
+from lightning.fabric.strategies import FSDPStrategy
+from lightning.fabric.utilities.throughput import Throughput, get_available_flops, measure_flops
+from lightning.fabric.loggers import CSVLogger
+from lightning.pytorch.loggers import WandbLogger
+from torch.utils.data import DataLoader
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_gpt.model import GPT, Block, Config, LLaMAMLP
+from lit_gpt.packed_dataset import CombinedDataset
+from lit_gpt.utils import chunked_cross_entropy, num_parameters
+
+# System settings
+model_name = "tiny-llama-1.1b"
+name = "lit-tiny-llama-1.1b"
+out_dir = Path("out") / name
+use_wandb = False
+
+# Hyperparameters
+devices = 4  # TODO: set this to 8
+
+global_batch_size = 32  # TODO: should be 512
+learning_rate = 4e-4
+micro_batch_size = 1  # TODO: should be 8
+max_step = 715256 * 2
+warmup_steps = 2000
+log_step_interval = 2
+eval_iters = 100
+save_step_interval = 1000
+eval_step_interval = 1000
+
+weight_decay = 1e-1
+beta1 = 0.9
+beta2 = 0.95
+grad_clip = 1.0
+decay_lr = True
+min_lr = 4e-5
+
+batch_size = global_batch_size // devices
+gradient_accumulation_steps = batch_size // micro_batch_size
+assert gradient_accumulation_steps > 0
+warmup_iters = warmup_steps * gradient_accumulation_steps
+
+max_iters = max_step * gradient_accumulation_steps
+lr_decay_iters = max_iters
+log_iter_interval = log_step_interval * gradient_accumulation_steps
+
+
+hparams = {k: v for k, v in locals().items() if isinstance(v, (int, float, str)) and not k.startswith("_")}
+
+
+def setup(resume: Union[bool, Path] = False):
+    if use_wandb:
+        logger = WandbLogger(project="tinyllama", name="training", resume=(resume is not False))
+    else:
+        logger = CSVLogger(root_dir="logs", name="tinyllama")
+
+    if devices > 1:
+        strategy = FSDPStrategy(
+            auto_wrap_policy={Block},
+            activation_checkpointing_policy=None,
+            state_dict_type="full",
+            limit_all_gathers=True,
+            cpu_offload=False,
+        )
+    else:
+        strategy = "auto"
+
+    fabric = L.Fabric(devices=devices, strategy=strategy, precision="bf16-mixed", loggers=[logger])
```

Comment:
On one hand everyone who wants to repeat the pretraining is gonna use modern GPUs that support `bf16` precision.
On the other - we have a function that automatically detects supported precision: either `bf16` or a regular `16`.
Could you use it here too?

Reply:
I want to set it like this for now so we can reproduce the same settings as in tinyllama. Ampere GPUs are pretty much a requirement here.

Reply:
Oh, ok.
I saw that this PR is no longer a `Draft`, so I decided that you are ready to ship it and thus started my nitpicking process 🤓.